users/views.py

```diff
@@ -0,0 +1,164 @@
+from django.db.utils import IntegrityError
+from django.http import Http404
+from django.shortcuts import get_object_or_404
+from rest_framework import status
+from rest_framework.authentication import TokenAuthentication
+from rest_framework.authtoken.models import Token
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from .models import User
+
+# pylint: disable= no-self-use, no-member
+
+
+class LoginUser(APIView):
+    """
+    This api request logs a user in and returns a token
+    """
+
+    def post(self, request):
+        """
+        This functions validates the credentials and logs a user in
+        Args:
+            request(HttpRequest): Value containing request data
+        Returns:
+            (Response): A json object containing message, code and token
+        """
+        response = {
+            "message": "Wrong email or password",
+            "code": status.HTTP_400_BAD_REQUEST,
+        }
+        try:
+            email = request.POST["email"]
+            password = request.POST["password"]
+            user = get_object_or_404(User, email=email)
+            if check_password(password, user.password):
+                response["message"] = "Logged in successfully"
+                response["code"] = status.HTTP_200_OK
+                if not Token.objects.filter(user=user).exists():
+                    token = Token.objects.create(user=user)
+                else:
+                    token = Token.objects.get(user=user)
+                response["token"] = token.key
+        except KeyError:
+            response["message"] = "Bad Request"
+        except Http404:
+            response["message"] = "User does not exist"
+
+        return Response(response)
+
+
+class AddUser(APIView):
+    """
+    This is a post request to add a user to db
+    """
+
+    def post(self, request):
+        """
+        This function adds a user to db
+        Args:
+            request(HttpRequest): Value containing request data
+        Returns:
+            (Response): A json object containing message and code
+        """
+        response = {}
+        try:
+            name = request.POST["name"]
+            email = request.POST["email"]
+            password = request.POST["password"]
+            validate_password(password)
+            password = make_password(password)
+            validate_email(email)
+            user = User(name=name, email=email, password=password)
+            user.save()
+            response["message"] = "successfully created user"
+            response["code"] = status.HTTP_201_CREATED
+        except KeyError:
+            response["message"] = "Error. Try again"
+            response["code"] = status.HTTP_400_BAD_REQUEST
+        except IntegrityError:
+            response["message"] = "Email already exists"
+            response["code"] = status.HTTP_409_CONFLICT
```

Comment:
`IntegrityError`  will not always mean that `Email` already exists. We should check the existing email upwards and throw the error from there. There will be multiple conditions to check for:
- Email exists : `User.objects.filter(email=<email>).exists()` What is the email in the post data is coming from an update user request, This should be case insensitive.
-

Reply:
Done